File: sdk/python/agent_scheduler_client.py
"""Agent Scheduler — Python client SDK.

A thin, dependency-light wrapper over the scheduler's REST API. Returns plain
dicts shaped like the API's ``JobView`` and raises typed errors so callers can
branch on failure mode rather than parsing status codes.

Requires: httpx (``pip install httpx``).

Example
-------
    from agent_scheduler_client import SchedulerClient

    with SchedulerClient("http://agent-scheduler-app:6816") as sched:
        sched.create_cron(
            job_id="nightly-report",
            cron_expression="0 2 * * *",
            event_data={"report": "daily_sales"},
        )
        for job in sched.list_jobs():
            print(job["job_id"], "->", job["resolved_stream"], job["next_run_time"])
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

import httpx

__all__ = [
    "SchedulerClient",
    "SchedulerError",
    "JobNotFoundError",
    "JobConflictError",
    "ValidationError",
    "ServiceUnavailableError",
]


# --- errors -----------------------------------------------------------------

class SchedulerError(Exception):
    """Base error. ``status`` is the HTTP code; ``detail`` the server message."""

    def __init__(self, status: int, detail: Any):
        self.status = status
        self.detail = detail
        super().__init__(f"[{status}] {detail}")


class JobNotFoundError(SchedulerError):
    """404 — no job with that id."""


class JobConflictError(SchedulerError):
    """409 — a job with that id already exists."""


class ValidationError(SchedulerError):
    """422 — invalid request body / trigger args."""


class ServiceUnavailableError(SchedulerError):
    """503 — scheduler is degraded (Valkey or job store unreachable)."""


_ERROR_BY_STATUS = {
    404: JobNotFoundError,
    409: JobConflictError,
    422: ValidationError,
    503: ServiceUnavailableError,
}


# --- client -----------------------------------------------------------------

class SchedulerClient:
    """Synchronous client for the Agent Scheduler admin API."""

    def __init__(
        self,
        base_url: str,
        *,
        timeout: float = 10.0,
        client: Optional[httpx.Client] = None,
    ):
        self._owns_client = client is None
        self._http = client or httpx.Client(
            base_url=base_url.rstrip("/"), timeout=timeout
        )

    # context manager sugar
    def __enter__(self) -> "SchedulerClient":
        return self

    def __exit__(self, *exc) -> None:
        self.close()

    def close(self) -> None:
        if self._owns_client:
            self._http.close()

    # --- low-level ----------------------------------------------------------

    def _request(self, method: str, path: str, **kwargs) -> Any:
        resp = self._http.request(method, path, **kwargs)
        if resp.is_success:
            if resp.status_code == 204 or not resp.content:
                return None
            return resp.json()
        # Error: unwrap FastAPI's {"detail": ...}
        try:
            detail = resp.json().get("detail", resp.text)
        except Exception:  # noqa: BLE001
            detail = resp.text
        raise _ERROR_BY_STATUS.get(resp.status_code, SchedulerError)(
            resp.status_code, detail
        )

    # --- ops ----------------------------------------------------------------

    def health(self) -> dict:
        """Liveness + Valkey/job-store reachability. Raises on 503."""
        return self._request("GET", "/health")

    def list_jobs(self) -> list[dict]:
        return self._request("GET", "/jobs")

    def get_job(self, job_id: str) -> dict:
        return self._request("GET", f"/jobs/{job_id}")

    def create_job(
        self,
        *,
        job_id: str,
        trigger_type: str,
        trigger_args: dict,
        target_stream_id: Optional[str] = None,
        event_type: str = "schedule.fired",
        event_data: Optional[dict] = None,
        room: Optional[str] = None,
        paused: bool = False,
    ) -> dict:
        """Create a job. ``trigger_type`` ∈ {interval, cron, date}; ``trigger_args``
        carries the fields for that type (see create_interval/cron/date helpers)."""
        body = {
            "job_id": job_id,
            "trigger_type": trigger_type,
            "trigger_args": trigger_args,
            "target_stream_id": target_stream_id,
            "event_type": event_type,
            "event_data": event_data or {},
            "room": room,
            "paused": paused,
        }
        return self._request("POST", "/jobs", json=body)

    # convenience constructors --------------------------------------------------

    def create_interval(
        self,
        job_id: str,
        *,
        seconds: int = 0,
        minutes: int = 0,
        hours: int = 0,
        days: int = 0,
        weeks: int = 0,
        **kwargs,
    ) -> dict:
        args = {
            k: v
            for k, v in dict(
                seconds=seconds, minutes=minutes, hours=hours, days=days, weeks=weeks
            ).items()
            if v
        }
        return self.create_job(
            job_id=job_id, trigger_type="interval", trigger_args=args, **kwargs
        )

    def create_cron(self, job_id: str, cron_expression: str, **kwargs) -> dict:
        return self.create_job(
            job_id=job_id,
            trigger_type="cron",
            trigger_args={"cron_expression": cron_expression},
            **kwargs,
        )

    def create_date(self, job_id: str, run_date: "str | datetime", **kwargs) -> dict:
        if isinstance(run_date, datetime):
            run_date = run_date.isoformat()
        return self.create_job(
            job_id=job_id,
            trigger_type="date",
            trigger_args={"run_date": run_date},
            **kwargs,
        )

    def update_job(
        self,
        job_id: str,
        *,
        trigger_type: Optional[str] = None,
        trigger_args: Optional[dict] = None,
        target_stream_id: Optional[str] = None,
        event_type: Optional[str] = None,
        event_data: Optional[dict] = None,
        room: Optional[str] = None,
    ) -> dict:
        """Partial update. Supply trigger_type + trigger_args together to reschedule."""
        body = {
            k: v
            for k, v in dict(
                trigger_type=trigger_type,
                trigger_args=trigger_args,
                target_stream_id=target_stream_id,
                event_type=event_type,
                event_data=event_data,
                room=room,
            ).items()
            if v is not None
        }
        return self._request("PATCH", f"/jobs/{job_id}", json=body)

    def delete_job(self, job_id: str) -> None:
        self._request("DELETE", f"/jobs/{job_id}")

    def pause_job(self, job_id: str) -> dict:
        return self._request("POST", f"/jobs/{job_id}/pause")

    def resume_job(self, job_id: str) -> dict:
        return self._request("POST", f"/jobs/{job_id}/resume")

    def run_job(self, job_id: str) -> dict:
        """Emit once now, off-schedule. Returns {status, job_id, entry_id}."""
        return self._request("POST", f"/jobs/{job_id}/run")

File: sdk/javascript/agentSchedulerClient.js
/**
 * Agent Scheduler — JavaScript (ES6) client SDK.
 *
 * Zero-dependency ES module over the scheduler's REST API. Uses the native
 * `fetch` (browser, Node 18+, Deno, Bun). Methods return Promises that resolve
 * to plain objects shaped like the API's `JobView`, and reject with a typed
 * `SchedulerError` subclass so callers can branch on `err.status`.
 *
 * @example
 *   import { SchedulerClient } from "./agentSchedulerClient.js";
 *   const sched = new SchedulerClient("http://agent-scheduler-app:6816");
 *   await sched.createCron("nightly-report", "0 2 * * *", {
 *     eventData: { report: "daily_sales" },
 *   });
 *   for (const job of await sched.listJobs()) {
 *     console.log(job.job_id, "->", job.resolved_stream, job.next_run_time);
 *   }
 */

export class SchedulerError extends Error {
  constructor(status, detail) {
    super(`[${status}] ${typeof detail === "string" ? detail : JSON.stringify(detail)}`);
    this.name = "SchedulerError";
    this.status = status;
    this.detail = detail;
  }
}
export class JobNotFoundError extends SchedulerError {}      // 404
export class JobConflictError extends SchedulerError {}      // 409
export class ValidationError extends SchedulerError {}       // 422
export class ServiceUnavailableError extends SchedulerError {} // 503

const ERROR_BY_STATUS = {
  404: JobNotFoundError,
  409: JobConflictError,
  422: ValidationError,
  503: ServiceUnavailableError,
};

export class SchedulerClient {
  /**
   * @param {string} baseUrl  e.g. "http://agent-scheduler-app:6816"
   * @param {object} [opts]
   * @param {number} [opts.timeoutMs=10000]
   * @param {typeof fetch} [opts.fetch]  inject a custom fetch (tests/SSR)
   */
  constructor(baseUrl, { timeoutMs = 10000, fetch: fetchImpl } = {}) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.timeoutMs = timeoutMs;
    this._fetch = fetchImpl || globalThis.fetch.bind(globalThis);
  }

  async _request(method, path, body) {
    const ctrl = new AbortController();
    const timer = setTimeout(() => ctrl.abort(), this.timeoutMs);
    let resp;
    try {
      resp = await this._fetch(this.baseUrl + path, {
        method,
        headers: body !== undefined ? { "Content-Type": "application/json" } : undefined,
        body: body !== undefined ? JSON.stringify(body) : undefined,
        signal: ctrl.signal,
      });
    } finally {
      clearTimeout(timer);
    }

    if (resp.ok) {
      if (resp.status === 204) return null;
      const text = await resp.text();
      return text ? JSON.parse(text) : null;
    }

    let detail;
    try {
      detail = (await resp.json()).detail;
    } catch {
      detail = await resp.text().catch(() => resp.statusText);
    }
    const Err = ERROR_BY_STATUS[resp.status] || SchedulerError;
    throw new Err(resp.status, detail);
  }

  // --- ops ------------------------------------------------------------------

  /** Liveness + Valkey/job-store reachability. Rejects on 503. */
  health() {
    return this._request("GET", "/health");
  }

  listJobs() {
    return this._request("GET", "/jobs");
  }

  getJob(jobId) {
    return this._request("GET", `/jobs/${encodeURIComponent(jobId)}`);
  }

  /**
   * Create a job.
   * @param {object} spec
   * @param {string} spec.jobId
   * @param {"interval"|"cron"|"date"} spec.triggerType
   * @param {object} spec.triggerArgs
   * @param {string|null} [spec.targetStreamId]
   * @param {string} [spec.eventType="schedule.fired"]
   * @param {object} [spec.eventData={}]
   * @param {string|null} [spec.room]
   * @param {boolean} [spec.paused=false]
   */
  createJob({
    jobId,
    triggerType,
    triggerArgs,
    targetStreamId = null,
    eventType = "schedule.fired",
    eventData = {},
    room = null,
    paused = false,
  }) {
    return this._request("POST", "/jobs", {
      job_id: jobId,
      trigger_type: triggerType,
      trigger_args: triggerArgs,
      target_stream_id: targetStreamId,
      event_type: eventType,
      event_data: eventData,
      room,
      paused,
    });
  }

  // convenience constructors --------------------------------------------------

  /** @param {object} interval e.g. { seconds: 300 } or { minutes: 15 } */
  createInterval(jobId, interval, opts = {}) {
    const args = Object.fromEntries(
      Object.entries(interval).filter(([, v]) => v)
    );
    return this.createJob({ jobId, triggerType: "interval", triggerArgs: args, ...opts });
  }

  createCron(jobId, cronExpression, opts = {}) {
    return this.createJob({
      jobId,
      triggerType: "cron",
      triggerArgs: { cron_expression: cronExpression },
      ...opts,
    });
  }

  /** @param {string|Date} runDate ISO-8601 string or Date */
  createDate(jobId, runDate, opts = {}) {
    const iso = runDate instanceof Date ? runDate.toISOString() : runDate;
    return this.createJob({
      jobId,
      triggerType: "date",
      triggerArgs: { run_date: iso },
      ...opts,
    });
  }

  /** Partial update. Pass triggerType + triggerArgs together to reschedule. */
  updateJob(jobId, fields = {}) {
    const map = {
      triggerType: "trigger_type",
      triggerArgs: "trigger_args",
      targetStreamId: "target_stream_id",
      eventType: "event_type",
      eventData: "event_data",
      room: "room",
    };
    const body = {};
    for (const [k, apiKey] of Object.entries(map)) {
      if (fields[k] !== undefined && fields[k] !== null) body[apiKey] = fields[k];
    }
    return this._request("PATCH", `/jobs/${encodeURIComponent(jobId)}`, body);
  }

  deleteJob(jobId) {
    return this._request("DELETE", `/jobs/${encodeURIComponent(jobId)}`);
  }

  pauseJob(jobId) {
    return this._request("POST", `/jobs/${encodeURIComponent(jobId)}/pause`);
  }

  resumeJob(jobId) {
    return this._request("POST", `/jobs/${encodeURIComponent(jobId)}/resume`);
  }

  /** Emit once now, off-schedule. Resolves to { status, job_id, entry_id }. */
  runJob(jobId) {
    return this._request("POST", `/jobs/${encodeURIComponent(jobId)}/run`);
  }
}
